core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1150,22 +1150,24 @@ sealed trait ZIO[-R, +E, +A] extends Serializable with ZIOPlatformSpecific[R, E,
    * in the background.
    */
   final def race[R1 <: R, E1 >: E, A1 >: A](that: ZIO[R1, E1, A1]): ZIO[R1, E1, A1] =
-    ZIO.fiberId
-      .flatMap(parentFiberId =>
-        (self raceWith that)(
-          (exit, right) =>
-            exit.foldM[Any, E1, A1](
-              cause => right.join mapErrorCause (cause && _),
-              a => (right interruptAs parentFiberId) as a
-            ),
-          (exit, left) =>
-            exit.foldM[Any, E1, A1](
-              cause => left.join mapErrorCause (_ && cause),
-              a => (left interruptAs parentFiberId) as a
-            )
-        )
+    ZIO.descriptorWith { descriptor =>
+      val parentFiberId   = descriptor.id
+      val uninterruptible = descriptor.interruptStatus.isUninterruptible
+      val left            = if (uninterruptible) self.disconnect else self
```

Comment:
You can use `self.uninterruptible.disconnect.interruptible` here, without changing interruptibility for all uses of disconnect